### ISSUE TITLE
where is the part of quantization? 

### ISSUE THREAD

**Issue by chenslcool** [AmanPriyanshu/FedPAQ-MNIST-implemenation#1]
Use np.around for quantization? Maybe it is different to  FedPAQ and doesn't work?

**Comment by AmanPriyanshu** [comment]
Yup, most definitely. FedPAQ, uses direct implementation from the paper "[Decentralized Stochastic Optimization and Gossip Algorithms with Compressed Communication](https://arxiv.org/pdf/1902.00340.pdf)". 

> Informally, we can think of Q as either a sparsification operator (that enforces sparsity of Q(g)) or a quantization operator that reduces the number of bits required to represent Q(g). For instance random rounding to less precise floating point numbers or to integers.

- From the same paper

So, instead of taking compression in the powers of 2 (usually implemented as such), I took to reduce the number of floating points, using np.around(). I did not take into account sparsification as well, which has been employed in both the papers.

Another deviation from paper implementation I made is that I quantized the gradient itself, rather than the difference, 
`Q(x(i)k,τ − xk)`. 

The repercussion of having done these are that:
1. The quantization/compression segment of the paper has not been completely implemented. Although I did implement quantization, sparsification has been dropped. The results will, therefore, be better than the actual implementation, as the sparsification uses a probability-based system to retain high-value points. But at the same time, it will contain more memory than the actual implementation, making it communicationally less viable.
2. Not using the difference `Q(x(i)k,τ − xk) `, makes it so that, if actually implemented, a larger data will have to be carried back and forth, i.e. it becomes communicationally less efficient, although both have been proven to converge in both these papers.

As for whether it works or not, it works, as the graphs are able to show a growing/diminishing curve as precision is increased/decreased respectively. But, it is not the actual implementation, as quantization would mean taking powers of 2, like 32-bit representation, to 16 bit. So, compared to that implementation, the effects of rounding are more prominent.

You can take a look at,


Floating Point Bitdepth | Largest value | Decimal digits of precision2
-- | -- | --
32-bit Float | 3.4028237 × 1038 | 7.22
16-bit Float | 6.55 × 104 | 3.31
14-bit Float | 6.55 × 104 | 3.01
11-bit Float | 6.50 × 104 | 2.1

So, I went from precision 7, 6, and 5. As they showed significant training loss, to make clean graphs.
